### ISSUE TITLE
Problem with SAEM covariate modeling

### ISSUE THREAD

**Issue by roninsightrx** [nlmixr2/nlmixr2#147]
hi Matt,

I'm having an issue when running SAEM in nlmixr2, hope you can help me.

I'm run a simple 2-cmt PK model and testing various covariates. The issue is that SAEM is not giving me sensible results,
basically ignoring any covariate I try to test and just sampling a dummy parameter (it seems)... 

Let's say I have the two models below, model 5 testing CRCL as a covariate, and model 6 testing WT. I'm using linear solved model here, but the same issue occurs with ODE systems.

```
mod5 <- function() {
  ini({
    tCL     <- log(5)
    tV      <- log(50)
    tQ      <- log(5)
    tV2     <- log(100)
    tCRCL_CL <- 0.5
    eta_CL   ~ 1.0
    eta_V    ~ 1.0
    add_sd <- 0.5
  })
  model({
    CL <- exp(tCL + eta_CL + logCRCL100*tCRCL_CL )
    V  <- exp(tV  + eta_V)
    Q  <- exp(tQ)
    V2 <- exp(tV2)
    linCmt() ~ add(add_sd)
  })
}

mod6 <- function() {
  ini({
    tCL     <- log(5)
    tV      <- log(50)
    tQ      <- log(5)
    tV2     <- log(100)
    tWT_CL  <- 0.5
    eta_CL  ~ 1.0
    eta_V   ~ 1.0
    add_sd <- 0.5
  })
  model({
    CL <- exp(tCL + eta_CL + logWT70*tWT_CL)
    V  <- exp(tV  + eta_V)
    Q  <- exp(tQ)
    V2 <- exp(tV2)
    linCmt() ~ add(add_sd)
  })
}
```
Note the different covariate estimate.
The head of my dataset looks like below. Obviously logCRCL100 and logWT70 are not the same. I can send you the full data via email if you like.
```
  ID    TIME   DV EVID MDV CMT  AMT RATE   logWT70  logFFM70 logCRCL100 caplogCRCL100       logCR   logCRstd
1  1  0.0000  0.0    1   1   1 2500 1000 0.6411017 0.1107744  0.6042130     0.4054651  0.00000000 -0.1457015
2  1  3.7000 27.9    0   0   1    0    0 0.6411017 0.1107744  0.6042130     0.4054651  0.00000000 -0.1457015
3  1 10.9167  8.6    0   0   1    0    0 0.6411017 0.1107744  0.6042130     0.4054651  0.00000000 -0.1457015
4  1 11.4500  0.0    1   1   1 1500 1000 0.6411017 0.1107744  0.6042130     0.4054651  0.00000000 -0.1457015
5  1 19.6000  0.0    1   1   1 1500 1000 0.6411017 0.1107744  0.6875946     0.4054651 -0.08338161 -0.1457015
6  1 27.2500  0.0    1   1   1 1500 1000 0.6411017 0.1107744  0.6875946     0.4054651 -0.08338161 -0.1457015
```

When I then sample:
```
res5 <- nlmixr(
  mod5, 
  data, 
  est = "saem", control = list(nBurn=50, nEm=50, print=25)
)
res6 <- nlmixr(
  mod6, 
  data, 
  est = "saem", control = list(nBurn=50, nEm=50, print=25)
)
```

The parameter estimates are exactly the same 
```
> res5$parFixed
          Est.     SE %RSE Back-transformed(95%CI) BSV(CV%) Shrink(SD)%
tCL       1.43  0.186 13.1       4.16 (2.89, 5.99)     72.3      3.02%<
tV        2.02 0.0382 1.89       7.57 (7.02, 8.16)     34.1      16.9%<
tQ        4.08 0.0686 1.68       59.2 (51.8, 67.7)                     
tV2      0.577  0.065 11.3       1.78 (1.57, 2.02)                     
tCRCL_CL  4.06 0.0429 1.06       4.06 (3.98, 4.14)                     
add_sd    2.83                                2.83                     
> res6$parFixed
        Est.     SE %RSE Back-transformed(95%CI) BSV(CV%) Shrink(SD)%
tCL     1.43  0.186 13.1       4.16 (2.89, 5.99)     72.3      3.02%<
tV      2.02 0.0382 1.89       7.57 (7.02, 8.16)     34.1      16.9%<
tQ      4.08 0.0686 1.68       59.2 (51.8, 67.7)                     
tV2    0.577  0.065 11.3       1.78 (1.57, 2.02)                     
tWT_CL  4.06 0.0429 1.06       4.06 (3.98, 4.14)                     
add_sd  2.83                                2.83          
```
One interesting thing, and perhaps related, is that the estimates are actually wrongly reported in the table here. CL is correct, but the order of the other estimates are all wrong.

Do you have any idea what is going on here?

**Comment by mattfidler** [comment]
My guess is there is some issue with either the `mu` referencing or how `nlmixr2` keeps the parameters.

You can try with 

```r
mod5 <- nlmixr(mod5, data, "saem", control=saemControl(muRefCov=FALSE))
```

If it gives reasonable results this is probably the case, and it needs to be fixed.

If the data are not proprietary, I would be OK taking it at my email `matt.fidler at novartis.com`

**Comment by mattfidler** [comment]
> When I then sample ... The parameter estimates are exactly the same

I'm not sure what you mean by "sampling"; do you mean that you expect the same saem to give different values since it is a `mcmc` simulation?  If so, that is not true, the seed is set to `99` every time you run `saem` so the results should be reproducible.  You could change the seed manually.

If you are resampling the data, then there likely should be some difference.

**Comment by roninsightrx** [comment]
> ```r
> muRefCov=FALSE
> ```
I tried, but it doesn't resolve the issue, still getting the same estimates for two different covariates. 
I did find one more nugget of information: I was running this on Ubuntu before, but now I tried running this on Windows, and I get the same behavior but now with additional error messages that might point to the problem... In the output object I'm getting this:

```
Information about run found (res5$runInfo):
  • column 'logCRCL100' has only 'NA' values for id '84' 
• column 'logCRCL100' has only 'NA' values for id '82' 
• column 'logCRCL100' has only 'NA' values for id '81' 
• column 'logCRCL100' has only 'NA' values for id '80' 
• column 'logCRCL100' has only 'NA' values for id '78' 
• column 'logCRCL100' has only 'NA' values for id '77' 
• column 'logCRCL100' has only 'NA' values for id '75' 
... etc for most (but not all) IDs in the dataset. 
```

So perhaps some parsing of the dataset by nlmixr is changing the covariate data? Or perhaps some indexing is wrong, also leading to the wrong order of parameters reported? I'll send you the dataset and script in a minute. Thanks for your time!

Yes, I know about the seed. I meant that I was expecting different results even with the similar seed because I'm testing two different covariate models. With "sampling" I meant that it's just ignoring the covariate altogether and having an unidentifiable parameter there.

**Comment by mattfidler** [comment]
> I tried, but it doesn't resolve the issue, still getting the same estimates for two different covariates. I did find one more nugget of information: I was running this on Ubuntu before, but now I tried running this on Windows, and I get the same behavior but now with additional error messages that might point to the problem... In the output object I'm getting this:

This is a false warning that I fixed

**Comment by mattfidler** [comment]
The internal datasets seem to be the different:

```r
  ID    TIME  EVID   AMT II   DV CMT logCRCL100 nlmixrRowNums
1  1  0.0000 10101  1000  0   NA   1   0.604213             1
2  1  2.5000 10101 -1000  0   NA   1         NA             0
3  1  3.7000     0    NA  0 27.9   1   0.604213             2
4  1 10.9167     0    NA  0  8.6   1   0.604213             3
5  1 11.4500 10101  1000  0   NA   1   0.604213             4
6  1 12.9500 10101 -1000  0   NA   1         NA             0

  ID    TIME  EVID   AMT II   DV CMT   logWT70 nlmixrRowNums
1  1  0.0000 10101  1000  0   NA   1 0.6411017             1
2  1  2.5000 10101 -1000  0   NA   1        NA             0
3  1  3.7000     0    NA  0 27.9   1 0.6411017             2
4  1 10.9167     0    NA  0  8.6   1 0.6411017             3
5  1 11.4500 10101  1000  0   NA   1 0.6411017             4
6  1 12.9500 10101 -1000  0   NA   1        NA             0
```

**Comment by roninsightrx** [comment]
I see, that seem right then.

When I put a breakpoint in nlmixr2est just before `.configsaem` is called, I'm seeing that my covariate is actually listed as a parameter. Is that correct?

<img width="376" alt="image" src="https://user-images.githubusercontent.com/15008273/236289911-98fb4107-b3a6-45f3-88db-6a54f28a6cf9.png">

**Comment by roninsightrx** [comment]
There is something wrong with the `.model` object. When I run an example that is working (based on your phenobarbital example, just adding weight as covariate), I see a `covars` attribute on `.model`. I don't see that when I run the above models. So seems something is wrong with `ui$saemModelList`.

**Comment by mattfidler** [comment]
> I see, that seem right then.
> 
> When I put a breakpoint in nlmixr2est just before `.configsaem` is called, I'm seeing that my covariate is actually listed as a parameter. Is that correct?
> 
> <img alt="image" width="376" src="https://user-images.githubusercontent.com/15008273/236289911-98fb4107-b3a6-45f3-88db-6a54f28a6cf9.png">

Yes, it is listed as a parameter.  Since it is time-varying it will not be a mu-referenced covariate.

**Comment by mattfidler** [comment]
> There is something wrong with the `.model` object. When I run an example that is working (based on your phenobarbital example, just adding weight as covariate), I see a `covars` attribute on `.model`. I don't see that when I run the above models. So seems something is wrong with `ui$saemModelList`.
> 
> It seems like in .saemFitModel it is removing time-varying covariates:
> 
> ```
> if (length(timeVaryingCovariates) > 0) {
>       # Drop time-varying covariates
>       .muRefCovariateDataFrame <- .muRefCovariateDataFrame[!(.muRefCovariateDataFrame$covariate %in% timeVaryingCovariates), ]
>     }
> ```
> 
> But I thought time-varying covariates were now possible in SAEM in nlmixr2?

Indeed it is possible.  Time-varying covariates are treated as Monolix-type regressor covariates. 

The time-varying covariates are not known until the data frame is provided; once it is known that the covariate is time varying it is removed from the mu-referenced parameter list.  This means that they are not part of the  the eventual `phi` parameter and will be included in the model. 

If you tried with the non time-varying WT, you would see that the covariate is no longer in the model at the same break-point.

**Comment by mattfidler** [comment]
> I see, that seem right then.
> 
> When I put a breakpoint in nlmixr2est just before `.configsaem` is called, I'm seeing that my covariate is actually listed as a parameter. Is that correct?
> 
> <img alt="image" width="376" src="https://user-images.githubusercontent.com/15008273/236289911-98fb4107-b3a6-45f3-88db-6a54f28a6cf9.png">

I was testing the data provided to the underlying `saem`.  I will also explore the data right before the configuration of `saem` to try to figure out why these are giving the same results.

**Comment by mattfidler** [comment]
You can see the underlying `rxode2()`  model by typing `summary(rxodeModel)`  So in this break-point I can get the same model, 

FYI this is the final model:

```R

Browse[3]> summary(attr(.model$saem_mod, "rx"))
rxode2 2.0.13.9000 model named rx_eef18c8317d114824689f56a90fe38b4 model (✔ ready). 
DLL: /tmp/RtmpI0Kb4r/rxode2/rx_eef18c8317d114824689f56a90fe38b4__.rxd/rx_eef18c8317d114824689f56a90fe38b4_.so
NULL

Calculated Variables:
[1] "rx_pred_"
── rxode2 Model Syntax ──
rxode2({
    param(tCL, tV, tQ, tV2, tCRCL_CL, logCRCL100)
    rx_pred_ = linCmtA(rx__PTR__, t, 0, 2, 1, exp(tCL + logCRCL100 * 
        tCRCL_CL), exp(tV), exp(tQ), exp(tV2), 0, 0, 0, 1, 0, 
        0, 0, 0, 1, 0, 0)
    cmt(rxLinCmt)
    dvid(1)
})
```

**Comment by mattfidler** [comment]
In this problem the data are listed as follows:

```
Browse[3]> head(.cfg$evt)
     ID    TIME  EVID   AMT II CMT LOGCRCL100
[1,]  0  0.0000 10101  1000  0   1   0.604213
[2,]  0  2.5000 10101 -1000  0   1         NA
[3,]  0  3.7000     0    NA  0   1   0.604213
[4,]  0 10.9167     0    NA  0   1   0.604213
[5,]  0 11.4500 10101  1000  0   1   0.604213
[6,]  0 12.9500 10101 -1000  0   1         NA
Browse[3]> head(.cfg$evtM)
  ID    TIME  EVID   AMT II CMT LOGCRCL100
1  0  0.0000 10101  1000  0   1   0.604213
2  0  2.5000 10101 -1000  0   1         NA
3  0  3.7000     0    NA  0   1   0.604213
4  0 10.9167     0    NA  0   1   0.604213
5  0 11.4500 10101  1000  0   1   0.604213
6  0 12.9500 10101 -1000  0   1         NA
```

This seems to match the data above, which means the data are relayed to this step appropriately.

**Comment by mattfidler** [comment]
It also correctly (internally) identifiers `logCRCL100` as a regressor or `inPars`:

```r
Browse[3]> attr(.model$saem_mod, "inPars")
[1] "logCRCL100
```

**Comment by mattfidler** [comment]
And after looking a bit more, the `logWT` is actually time-varying too, so all the above apply with a different dataset

**Comment by mattfidler** [comment]
You can also look at the parameter distributions of logwt and logcrcl and see that they are quite similar and with a lazy random walk like saem it is not impossible to get the same parameters with very similar but not identical distributions.

**Comment by roninsightrx** [comment]
> You can also look at the parameter distributions of logwt and logcrcl and see that they are quite similar and with a lazy random walk like saem it is not impossible to get the same parameters with very similar but not identical distributions.

Yes, I had considered (CRCL is calculated partially based on WT), but there is definitely some additional unexplained variability between the two covariates. Also the standard deviation for CRCL is almost 3x that of WT (after dividing by the mean and logging), so it would be surprising to me that SAEM would estimate the effects to be exactly the same. Also when destroying the relationship (`data = data %>% mutate(logWT70 = rev(logWT70))`), the parameter estimates are still exactly the same, leading me to believe that the covariate is just ignored when mu-modeled.

But thanks for suggesting to run with the time-varying covariates non-MU modeled, I will try that tomorrow for various covariates and report back.

**Comment by mattfidler** [comment]
My guess is the acceptance of the covariate parameter is the same between the two covariance during the mcmc/em walk, but increasing the variance will give different acceptences of the covariate estimates.

**Comment by roninsightrx** [comment]
Thanks Matt! When putting the covariate effect outside of the exponential, I'm getting very credible results. I think that solved it.

Should perhaps an error be thrown for future users that attempt to put time-varying covariates within the exponential in SAEM? Or do you think there could be cases where that will work fine and is still preferred? Even if so, a warning would still be in order probably? If not looking closely at the output a user might easily accept erroneous results.

The above also solved the issue with the parameter estimates getting shuffled in the output. Do you want me to make a separate ticket for that?

**Comment by mattfidler** [comment]
> Should perhaps an error be thrown for future users that attempt to put time-varying covariates within the exponential in SAEM? Or do you think there could be cases where that will work fine and is still preferred? Even if so, a warning would still be in order probably? If not looking closely at the output a user might easily accept erroneous results.

I think there may be use cases, but you have a valid point here; perhaps a warning would be helpful (at the very least).  It would probably be applicable for any log-transformed parameters.

> The above also solved the issue with the parameter estimates getting shuffled in the output. Do you want me to make a separate ticket for that?

This may have been because you were using an old version of nlmixr2.  I didn't actually have this issue when I tried your code.  If you can reproduce it with a recent version, then please.  Otherwise I am assuming it is fixed in the recent version.

**Comment by mattfidler** [comment]
Another option is to do some nealder-meade optimization for these parameters instead of only using the `mcmc` algorithm.  While it is not described in the methodology section of monolix, it could also work a bit better.
